File: src/test/testData/java/LogicalOperations.java
class Tests {
    @Complexity(4)
    public String simpleStatements() {
        if (true)                      // +1 if
            return "no conditions"

        if (a == b) {                  // +1 if
            return "simple equal"
        }

        if (                           // +1 if
            a == b && b == d) {        // +1 AND group
            return "still simple"
        }
        return "exit"
    }

    @Complexity(2)
    public void simpleAnd() {
        if (                           // +1 if
            a && b)                    // +1 AND
            return
    }


    @Complexity(2)
    public void simpleOr() {
        if (                           // +1 if
            a || b) {                  // +1 OR
            return
        }
    }


    @Complexity(2)
    public void singleLongGroup() {
        if (                          // +1 if
            a || b || c || d) {       // +1 OR GROUP
            return
        }
    }

    @Complexity(3)
    public void twoGroups() {
        if (                          // +1 if
            a || b ||                 // +1 OR
                c && d) {             // +1 AND
            return
        }
    }


    @Complexity(3)
    public void parenthesisCreateNewGroupAnyway() {
        if (                          // +1 if
            a || b ||                 // +1 OR
                (c || d)) {           // +1 OR separate
            return
        }
    }

    @Complexity(4)
    public void parenthesisInCenterSplitTheGroup() {
        if (                          // +1 if
            a || b ||                 // +1 OR
                !(c || d)             // +1 OR separate
                    || e || f) {      // +1 new OR
            return
        }
    }

    @Complexity(1)
    public boolean doesSupportOperation() {
        return exists && support(op)
    }
}

File: src/test/testData/go/logical_operations.go
package main

//go:generate complexity 7
func complex(
	a, b, c, d, e, f, g, h, i, j bool,
) bool {
	if        // +1
	a || b && // +1
		c || d && // +1
		e || f && // +1
		g || h && // +1
		i || j { // +1
		return true
	} else { // +1
		return false
	}
}

//go:generate complexity 4
func simpleStatements(b bool, v int) int {
	if b { // +1
		return 1
	}
	if v < 0 { // +1
		return 2
	}
	if v%3 == 0 && // +1
		v%5 == 0 { // +1
		return 15
	}
	return v
}

//go:generate complexity 3
func twoGroups(a, b, c, d bool) bool {
	if        // +1
	a || b || // +1 OR
		!(c || d) { // +1 OR separate
		return true
	}
	return false
}

type Aa struct {
	a, b, c, d bool
}

//go:generate complexity 3
func (a Aa) parenthesisCreateNewGroupAnyway() bool {
	if            //+1 if
	a.a || a.b || //+1 OR
		(a.c || a.d) { // +1 OR group
		return true
	}
	return false
}

//go:generate complexity 4
func (a Aa) parenthesisInCenterSplitTheGroup(e, f bool) bool {
	if            //+1 if
	a.a || a.b || //+1 OR
		!(a.c || // +1 OR group
			a.d) || e || f { // +1 new OR
		return true
	}
	return false
}

func (a Aa) call() bool {
	return true
}

//go:generate complexity 1
func (a Aa) doesSupportOperation() bool {
	return a.a && a.call()
}

//go:generate complexity 2
func (a Aa) doesSupportOperationInStatement() bool {
	r := a.a && a.call() //+1
	r = r && a.call()    //+1
	return r
}

type MayErr interface {
	runAction() error
	getMap() map[string]string
	consumeSomething(v string)
}

//go:generate complexity 1
func goErrorIfStatement(m MayErr) error {
	if err := m.runAction(); err != nil { // +1
		return err
	}
	return nil
}

//go:generate complexity 1
func goSpecificIfStatement(k string, m MayErr) {
	f := m.getMap()
	if v, ok := f[k]; ok { // +1
		m.consumeSomething(v)
	}
}
